### DISCUSSION TITLE
CoApp Not Recognized in LibreWolf - Linux Flatpak

### DISCUSSION THREAD

**Discussion by ufgkirk** [aclap-dev/video-downloadhelper#1131] [Help and Q&A]
### Your question

I just installed the CoApp on my Linux system, and it is working in Firefox, but not in LibreWolf.  Based on the output of the installer, it looks like the registration of LibreWolf was just skipped by the installer.

LibreWolf was installed from a Flatpak.  I am running a KDE Plasma environment on Debian 12.

Browser version:

> $ flatpak list | grep LibreWolf
> LibreWolf       io.gitlab.librewolf-community   128.0.3-2       stable  system

OS Version:

> $ lsb_release -a
> No LSB modules are available.
> Distributor ID: Debian
> Description:    Debian GNU/Linux 12 (bookworm)
> Release:        12
> Codename:       bookworm
> $ uname -a
> Linux Enterprise 6.1.0-23-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.99-1 (2024-07-15) x86_64 GNU/Linux

Installation output:

> $ curl -sSLf https://github.com/aclap-dev/vdhcoapp/releases/latest/download/install.sh | bash
> Downloading: https://github.com/aclap-dev/vdhcoapp/releases/latest/download/vdhcoapp-linux-x86_64.tar.bz2
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100 52.8M  100 52.8M    0     0  73.2M      0 --:--:-- --:--:-- --:--:-- 73.2M
> Extracting tarball…
> /tmp/vdhcoapp-ggzHSE: bzip2 compressed data, block size = 900k
> Registering CoApp
> Installing…
> Flatpak is installed. Making the coapp available from browser sandboxes:
> Linked coapp within org.mozilla.firefox.
> Linked coapp within com.brave.Browser.
> Linked coapp within com.google.Chrome.
> Linked coapp within com.google.ChromeDev.
> Linked coapp within org.chromium.Chromium.
> Linked coapp within com.github.Eloston.UngoogledChromium.
> Linked coapp within com.microsoft.Edge.
> Linked coapp within com.microsoft.EdgeDev.
> Writing /home/kirk/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json
> Writing /home/kirk/.var/app/org.mozilla.firefox/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json
> Writing /home/kirk/.var/app/com.brave.Browser/config/BraveSoftware/Brave-Browser/NativeMessagingHosts/net.downloadhelper.coapp.json
> Writing /home/kirk/.var/app/com.google.Chrome/config/google-chrome/NativeMessagingHosts/net.downloadhelper.coapp.json
> Writing /home/kirk/.var/app/com.microsoft.Edge/config/microsoft-edge/NativeMessagingHosts/net.downloadhelper.coapp.json
> VdhCoApp : VdhCoApp is ready to be used
> CoApp successfuly installed under '~/.local/share/vdhcoapp'.
> To uninstall, run '~/.local/share/vdhcoapp/vdhcoapp uninstall' and remove '~/.local/share/vdhcoapp'.
> Re-run that script to update the coapp.

Based on some discussion I found in [another post here about this issue on Mac OS](https://github.com/aclap-dev/video-downloadhelper/discussions/216), I tried copying the native-messaging-hosts folder from the Firefox directory to the LibreWolf directory, but this had no effect for me.  

> ~/.var/app$ ls org.mozilla.firefox/.mozilla/native-messaging-hosts/
> net.downloadhelper.coapp.json
> ~/.var/app$ cp -R org.mozilla.firefox/.mozilla/native-messaging-hosts io.gitlab.librewolf-community/.librewolf/
> $ ls io.gitlab.librewolf-community/.librewolf/native-messaging-hosts/
> net.downloadhelper.coapp.json

Vdhcoapp info dump:

> $ ~/.local/share/vdhcoapp/vdhcoapp --info
> {
>   "id": "net.downloadhelper.coapp",
>   "name": "VdhCoApp",
>   "version": "2.0.19",
>   "binary": "/home/kirk/.local/share/vdhcoapp/vdhcoapp",
>   "displayName": "VdhCoApp",
>   "description": "Video DownloadHelper companion app",
>   "target": {
>     "os": "linux",
>     "arch": "x86_64",
>     "node": "18"
>   },
>   "home": "/home/kirk",
>   "converterBinary": "/home/kirk/.local/share/vdhcoapp/ffmpeg",
>   "converterBase": "ffmpeg",
>   "converterBaseVersion": "c30f360"
> }

Hopefully it will be an easy fix, and maybe the installers can be updated to support LibreWolf in the future.  Thanks!

### Addon & CoApp version and details

version: 9.1.0.42
target: mozilla
channel: stable
lang: en-US
coapp: {"found":false,"error":"An unexpected error occurred"}
license: {"unneeded":true}
platform: x86-64 linux
UA: Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0
{
    "used_history_button": true,
    "never_show_no_incognito_msg_again": true,
    "view_options": {
        "all_tabs": false,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "hide_downloaded": false
    },
    "open_count_store": 4,
    "successfull_dl": 16
}

**Comment by Mattmjs** [comment]
Do you recall seeing a popup in the browser to give it permission to access the companion app. I have a screenshot of it for the
Firefox snap app on Ubuntu. You can see that it popped up after I opened the settings in the extension.

![ubuntu vm](https://github.com/user-attachments/assets/7fbc2f90-1171-4863-9432-f0144b649f99)

**Reply by ufgkirk** [reply]
> Do you recall seeing a popup in the browser to give it permission to access the companion app. I have a screenshot of it for the Firefox snap app on Ubuntu. You can see that it popped up after I opened the settings in the extension.

Hmm.... unfortunately no.

**Comment by paulrouget** [comment]
https://github.com/aclap-dev/video-downloadhelper/discussions/216 and maybe this:

```
flatpak permission-set webextensions net.downloadhelper.coapp io.gitlab.librewolf-community yes
```

**Reply by ufgkirk** [reply]
> #216 and maybe this:
> 
> ```
> flatpak permission-set webextensions net.downloadhelper.coapp io.gitlab.librewolf-community yes
> ```

I gave this a try, but it seems to have had no effect.

I actually did try the recommended solution in that other thread of copying folders from my .mozilla folder for my Firefox install into the .librewolf folder.  It had no effect either. lol

**Comment by bitsByB** [comment]
same problem for firefox here too...

**Comment by infinitieunique** [comment]
It seems like the default installer puts a `net.downloadhelper.coapp.json` files to the directories of almost all mainstream browsers and Librewolf isn't one of them. The solution is to copy the file into librewolf's folder manually to make it work:
`/home/user/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json`
`/home/user/.librewolf/native-messaging-hosts/net.downloadhelper.coapp.json`

![img-2024-11-21-17-13-17](https://github.com/user-attachments/assets/6db521e2-765f-4daa-9178-51802723fed8)

**Comment by bobafetthotmail** [comment]
the `flatpak permission-set webextensions net.downloadhelper.coapp io.gitlab.librewolf-community yes
` seems to have no effect at all even given as root (my flatpaks are installed system wide as root, as distro default, so if I install or update or change settings I have to use sudo, they run as user of course)

but I found a way to make it work using Flatseal flatpak application (or KDE flatpak application permission applet, which seems to operate the same settings). Flatseal is highly recommended if you use Flatpak so you can actually lock down all the Flatpak applications that ship with broken sandboxing options like full filesystem access or add access to custom folders in an application's sandbox like we are doing here.

1. download and extract the tar.gz in a generic folder that has no versions in the name (will be convenient when updating)

`/home/alby/vdhcoapp/`

2. run the install as user (NO ROOT NO SUDO)

`/home/alby/vdhcoapp/vdhcoapp install`

3. create the folder where we will move the file
`mkdir /home/alby/.var/app/io.gitlab.librewolf-community/.librewolf/native-messaging-hosts -p
`
4. copy file over from the one it installed for mozilla firefox
`cp /home/alby/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json /home/alby/.var/app/io.gitlab.librewolf-community/.librewolf/native-messaging-hosts/net.downloadhelper.coapp.json`

5. and now the missing step to make it all work. Use Flatseal to allow access to the vdhcoapp folder from librewolf's sandbox, in the Filesystem -- Other Files section
`/home/alby/vdhcoapp:ro`
the ":ro" at the end means this folder will be accessible readonly, for more security. It does not need to write there.



in theory Flatseal and KDE flatpak app permission gui do the same thing as doing in the terminal
`flatpak override io.gitlab.librewolf-community --filesystem=/home/alby/vhdcoapp:ro`
(run as root/sudo or user depending on how you installed flatpak)

but I don't see their overrides from commandline so I don't know where they store the settings.

I think/hope any future update to the companion app can just be done manually copying the binaries into the same folder overwriting the old ones. I have done that many times in the past before using librewolf and flatpak.
This way you don't need to repeat all the instructions for every update.

**Reply by bitsByB** [reply]
this is great, thanks a lot

**Reply by rgarrigue** [reply]
I installed librewolf through their apt repository, and CoApp with the script, and the path were different.

```bash
mkdir -p $HOME/.librewolf/native-messaging-hosts
cp $HOME/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json $HOME/.librewolf/native-messaging-hosts/net.downloadhelper.coapp.json
```

**Reply by nehemiagurl** [reply]
that was it, thank you!
though it is important to note that the install script installs the app at `~/.local/share/vdhcoapp` so step 5 needs to be done with that path instead.
